### PR TITLE
feat(optimizer): allow temporal filter in on clause

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
- add on clause in binder context
- ban the now expr in stream join executor

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/input/temporal_filter.yaml [hunk 1]
```diff
@@ -29,3 +29,45 @@
...
   - stream_plan
   - stream_dist_plan
+- name: Temporal filter in on clause 1
+  sql: |
+    create table t1 (a int, ta timestamp with time zone);
+    create table t2 (b int, tb timestamp with time zone);
+    select * from t1 join t2 on a = b AND ta < now() - interval '1 hour' and ta >= now() - interval '2 hour';
+  expected_outputs:
+  - stream_plan
+- name: Temporal filter in on clause 2
```
Comment (xxchan) [inline src/frontend/planner_test/tests/testdata/input/temporal_filter.yaml:39]: Descriptive names are better. e.g., "left join, ta"

File: src/frontend/src/binder/expr/function.rs [hunk 2]
```diff
@@ -949,7 +950,7 @@ impl Binder {
         if self.is_for_stream()
             && !matches!(
                 self.context.clause,
-                Some(Clause::Where) | Some(Clause::Having)
+                Some(Clause::Where) | Some(Clause::Having) | Some(Clause::On)
```
Comment (xxchan) [inline src/frontend/src/binder/expr/function.rs:953]: image

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 3]
```diff
@@ -1232,6 +1232,15 @@ impl ToBatch for LogicalJoin {
 
 impl ToStream for LogicalJoin {
     fn to_stream(&self, ctx: &mut ToStreamContext) -> Result<PlanRef> {
+        if self
+            .on()
+            .conjunctions
+            .iter()
+            .any(|cond| cond.count_nows() > 0)
+        {
+            return Err(ErrorCode::NotSupported("now() expression in ON clause".to_string(), "please refer to https://www.risingwave.dev/docs/current/sql-pattern-temporal-filters/ for more information".to_string()).into());
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:1241]: Can you explain more about why it's invalid? like https://github.com/risingwavelabs/risingwave/blob/103652136437d41812fb9482c70ad429f18a0c4c/src/frontend/src/optimizer/plan_node/logical_filter.rs#L204-L218

**Review by xxchan — APPROVED** [review]
generally lgtm

File: src/frontend/src/binder/bind_context.rs [hunk 4]
```diff
@@ -50,6 +50,7 @@ pub enum Clause {
     Where,
     Values,
     GroupBy,
+    On,
```
Comment (xiangjinwu) [inline src/frontend/src/binder/bind_context.rs:53]: Maybe call this `JoinOn`?

Comment (st1page) [reply src/frontend/src/binder/bind_context.rs:53]: Maybe, I will change `Filter` to `AggFilter` after I make sure it is that semantics :thinking:

**Pull request merged**